### ISSUE TITLE
Compile at server

### ISSUE THREAD

**Issue by benjamine** [alanshaw/grunt-template-client#2]
I noticed this plugin is rendering calls to `Hogan.compile`, that means templates will be compiled in the browser after downloading them.
I see that works, but the whole idea of precompilation is actually compile the templates, on server, and render the precompiled js functions in the module.
That would be a nice upgrade.

**Comment by alanshaw** [comment]
Not sure I understand what you mean. The output from the task _could_ be run on the server if you really wanted.

I don't know how the task would generate compiled templates that could be sent to the client - the compiled templates will be a javascript function or object which won't be serializable to send.

Once grunt has generated the compiled templates, what do you suppose it should do with them?

Either way, this plugin was forked from https://github.com/ullmark/grunt-hogan-client _but_ no longer deals specifically with Hogan so I'm not sure the issue is even relevant. As an aside, for it to be able to compile _any_ templates the plugin would have to depend on compilers for all current and future template engines - something I have no desire to maintain.

**Comment by benjamine** [comment]
Ok, let me explain a little before closing this,

I did the same comment at [grunt-hogan-client](https://github.com/ullmark/grunt-hogan-client), it has the same problem.
I like the idea of making it generic, server-side compilation could be done generically too.
Hogan, and others have the capability to compile a template into a serializable function (this is what makes server-side compilation possible).
As an example you can look at [grunt-hogan](https://github.com/automatonic/grunt-hogan) which got this correctly.

if you do this:

``` js
Hogan.compile('please {{compile}} me');
```

output is a javascript function that's useful for immediate execution but can't be sent to client, but if you do this:

``` js
Hogan.compile('please {{compile}} me', { asString: true });
```

the output is this string:

``` js
'function(c,p,i){var _=this;_.b(i=i||"");_.b("please ");_.b(_.v(_.f("compile",c,p,0)));_.b(" me");return _.fl();;}'
```

which can be safely print in a .js file and send to the browser if you just wrap it in `new Hogan.Template( ... )`

**Comment by alanshaw** [comment]
My bad. I didn't realise you could do this. I'll add this in. Shouldn't be too hard.

**Comment by alanshaw** [comment]
You can compile templates by passing a function as the `val` option. It receives the template content as an argument and should return a string: https://github.com/alanshaw/grunt-template-client#optionsval

Released to NPM as grunt-template-client@1.0.0

**Comment by benjamine** [comment]
:+1: sure, looks nice, glad to help
